design/sync-labels-bmh-to-node.md

```diff
@@ -0,0 +1,191 @@
+
+### Annotations on BMH
+
+In this approach, an annotation is added to a BMH, in addition to the labels to be synchronized, which reflects the prefixes to be utilized. For example,
+
+```
+  kind: BareMetalHost
+  metadata:
+    annotations:   
+      cluster.x-k8s.io/label-prefixes: "my-prefix.metal3.io, my-prefix.kubernetes.io"
+      cluster.x-k8s.io/label-sync-interval: "30s"   
+    labels:
+      my-prefix.metal3.io/rack: xyz-123
+      my-prefix.kubernetes.io/zone: security-level-0   
+```
+
+A concern with this approach is consistency since the same annotation would have to be duplicated across BMHs. The prefixes, while user configurable, are static and should not require re-configuring. For example, in a likely scenario, the cluster admin would define a standard set of prefixes to be used internally. Also, annotations can be (accidentally) added/removed/changed, leading to undesirable behavior. 
+
+### New API Type
+
+We could introduce a new API type: `BareMetalHostLabelSyncProfile` (any other suggestions for the name?).
+
+The common case is when a label on the BMH is updated, we must first find the corresponding `BareMetalHostLabelSyncProfile` (if any) that matches the BMH. To avoid fetching all `BareMetalHostLabelSyncProfile`, we can make use of a label on the BMH, in namespace/name format, that references the `BareMetalHostLabelSyncProfile` (see example below). We can drop the namespace requirement if it's in the same namespace as the BMH.
+
+The less common case is where the `BareMetalHostLabelSyncProfile` is updated (e.g. a prefix is removed). In this case, we will need to fetch all BMHs associated and perform a sync operations across the impacted Nodes. Label selectors could be used for this purpose. 
+
+The approach is illustrated below:
+
+```
+kind: BareMetalHost
+metadata:
+  labels:
+    metal3.io/baremetalhost-label-sync-profile: some-ns/label-sync-profile-0
+...
+---
+kind: BareMetalHostLabelSyncProfile
+name: label-sync-profile-0
+namespace: some-ns
+labelSelector: 
+  matchLabels: 
+    metal3.io/baremetalhost-label-sync-profile: some-ns/label-sync-profile-0
+prefixes:
+  - my-prefix.metal3.io
+  - my-prefix.kubernetes.io
+label-sync-interval: "30s"  
+```
+
+It's the admin's/user's responsibility to create the linkage by labeling BMHs (e.g `metal3.io/baremetalhost-label-sync-profile: some-ns/label-sync-profile-0`). 
+
+### Metal3Cluster 
+
+We can have the prefixes specified at the Metal3Cluster level? For example, 
+
+```
+kind: Metal3Cluster
+name: test1
+spec:
+  bmh-label-sync-prefixes: "my-prefix.metal3.io, my-prefix.kubernetes.io"
+  bmh-label-sync-interval: "30s"   
+...
+```
+
+We need to map the BMH to the Metal3Cluster object. Similarly, when Metal3Cluster is updated, we need to fetch all associated BMHs. We can investigate this further if this is a better approach. 
```

Comment:
This is already partially done, when we reconcile the metal3machine we fetch all objects including Cluster and Metal3cluster : https://github.com/metal3-io/cluster-api-provider-metal3/blob/master/controllers/metal3machine_controller.go#L110 and https://github.com/metal3-io/cluster-api-provider-metal3/blob/master/controllers/metal3machine_controller.go#L137. The reverse is done here : https://github.com/metal3-io/cluster-api-provider-metal3/blob/master/controllers/metal3machine_controller.go#L368 . It is quite straightforward to adapt to get the link with the BMH. IMO this approach is the best, considering that it does not require additional objects and controls the behaviour per cluster.